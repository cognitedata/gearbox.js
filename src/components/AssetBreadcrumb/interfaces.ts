import { Asset, CogniteInternalId } from '@cognite/sdk';

export type FetchAssetCall = (assetId: CogniteInternalId) => Promise<Asset>;

export interface AssetBreadcrumbProps {
  /**
   * Asset ID
   */
  assetId: number;
  /**
   * Maximal number of assets to be displayed.
   * If length of the asset chain bigger than < maxLength > value,
   * asset chain will be shrunk to root element plus < maxLength > - 1 number
   * of last elements in a chain
   */
  maxLength?: number;
  /**
   * Function, that can be used for custom rendering of asset in a breadcrumb
   */
  renderItem?: (asset: Asset, depth: number) => JSX.Element;
  /**
   * Function that can be used to replace embedded asset fetching
   */
  retrieveAsset?: FetchAssetCall;
  /**
   * Callback which is executed on a click action on asset in a breadcrumb.
   * It's only available in case of default asset rendering in a breadcrumb
   */
  onBreadcrumbClick?: (asset: Asset, depth: number) => void;
}
