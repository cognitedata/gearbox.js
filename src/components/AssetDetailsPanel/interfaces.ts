import { CSSProperties } from 'react';
import { WithAssetDataProps, WithAssetProps } from '../../hoc';
import { MetaDescriptionListProps } from '../DescriptionList';

export interface AssetDetailsPanelStylesProps {
  /**
   * Object that defines inline CSS style for container of the table
   */
  styles?: CSSProperties;
}

export type AssetDetailsPanelPureProps = WithAssetDataProps &
  MetaDescriptionListProps &
  AssetDetailsPanelStylesProps;

export type AssetDetailsPanelProps = WithAssetProps &
  MetaDescriptionListProps &
  AssetDetailsPanelStylesProps;
