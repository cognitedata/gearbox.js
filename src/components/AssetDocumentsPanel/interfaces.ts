import { FilesMetadata } from '@cognite/sdk';
import { CollapseProps } from 'antd/lib/collapse';
import { CSSProperties, ReactNode } from 'react';
import { WithAssetFilesDataProps, WithAssetFilesProps } from '../../hoc';
import { AnyIfEmpty } from '../../interfaces';

export type Document = FilesMetadata;

export type DocumentRenderer = (
  document: Document,
  i: number,
  documents: Document[]
) => ReactNode;

export type OnDocumentClick = (
  document: Document,
  category: string,
  description: string
) => void;

export interface JsonDocTypes {
  [s: string]: string;
}

export interface DocumentType {
  key: string;
  description: string;
}

export interface DocumentsByCategory {
  [s: string]: {
    description: string;
    documents: Document[];
  };
}

export interface Priority {
  [s: string]: number;
}

export interface DocumentsPanelStylesProps {
  /**
   * Object that defines inline CSS styles for inner elements of the component
   */
  styles?: AssetDocumentsPanelStyles;
}

export interface DocumentsPanelThemeProps {
  theme?: AnyIfEmpty<{}>;
}

export type DocumentTableProps = MetaDocProps &
  WithAssetFilesDataProps &
  DocumentsPanelStylesProps &
  DocumentsPanelThemeProps;

export interface AssetDocumentsPanelStyles {
  wrapper?: CSSProperties;
  fileContainer?: CSSProperties;
  fileLink?: CSSProperties;
  fileTitle?: CSSProperties;
}

export interface MetaDocProps {
  /**
   * Callback function triggered when user clicks on a file (document)
   */
  handleDocumentClick?: OnDocumentClick;
  /**
   * Object with props to be passed to atn design Collapse component
   */
  collapseProps?: CollapseProps;
  /**
   * List of categories codes to be shown on the top of the list.
   * Categories "P&ID" and "Logic Diagrams" are prioritized by default.
   */
  categoryPriorityList?: string[];
  /**
   * Name for the category that includes all files with undefined category
   */
  unknownCategoryName?: string;
  /**
   * Key to get document title from `metadata` object
   */
  documentTitleField?: string;
  /**
   * The `metadata` field used to group documents into types. If this is
   * not specified, then `doc_type` field will be attempted. If no
   * `doc_type` field is present in the metadata, then the filename will
   * be attempted to be parsed using the NORSOK standard. If this fails, then
   * the document will be in the "unknown" group.
   */
  documentTypeField?: string;
  /**
   * Object map used to show custom category names
   */
  docTypes?: JsonDocTypes;
  /**
   * Text to be shown if no documents have been found for the asset
   */
  noDocumentsSign?: string;
  /**
   * Custom render function to modify documents appearing
   */
  documentRenderer?: DocumentRenderer;
  /**
   * Function to sort categories after theirs prioritization
   */
  customCategorySort?: (a: string, b: string) => number;
}

export type AssetDocumentsPanelProps = WithAssetFilesProps &
  MetaDocProps &
  DocumentsPanelStylesProps;
