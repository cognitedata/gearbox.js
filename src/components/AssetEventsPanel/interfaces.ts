import { CSSProperties } from 'react';
import { WithAssetEventsDataProps, WithAssetEventsProps } from '../../hoc';
import { AnyIfEmpty } from '../../interfaces';

export interface TableColumnType {
  title: string;
  dataIndex: string;
  key?: string;
}

export interface TableDesignType {
  /**
   * Configure table pagination appearance
   */
  pagination?: {
    pageSize?: number;
    position?: 'top' | 'bottom' | 'both';
    showSizeChanger?: boolean;
  };
  /**
   * Configure scroll appearance
   */
  scroll?: {
    y?: string;
    x?: string;
  };
  /**
   * Define if table has borders
   */
  bordered?: boolean;
  /**
   * Toggles table header visibility
   */
  showHeader?: boolean;
  /**
   * Custom table styles
   */
  style?: CSSProperties;
}

export interface AssetEventsPanelStyles {
  table?: CSSProperties;
  tableRow?: CSSProperties;
  tableCell?: CSSProperties;
}

export interface MetaEventsProps extends TableDesignType {
  /**
   * Array of objects that customize titles of the columns in the table
   */
  columns?: TableColumnType[];
}

export interface AssetEventsPanelStylesProps {
  /**
   * Object that defines inline CSS styles for inner elements of the component
   */
  styles?: AssetEventsPanelStyles;
}

export interface AssetEventsPanelThemeProps {
  /**
   * @ignore
   */
  theme?: AnyIfEmpty<{}>;
}

export type AssetEventsPanelProps = WithAssetEventsProps &
  MetaEventsProps &
  AssetEventsPanelStylesProps &
  AssetEventsPanelThemeProps;

export type AssetEventsPanelPropsPure = WithAssetEventsDataProps &
  MetaEventsProps &
  AssetEventsPanelStylesProps &
  AssetEventsPanelThemeProps;
