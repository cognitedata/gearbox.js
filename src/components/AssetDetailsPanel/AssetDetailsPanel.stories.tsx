import React from 'react';
import { storiesOf } from '@storybook/react';
import AssetDetailsPanel, {
  AssetDetailsColumns,
} from 'components/AssetDetailsPanel/AssetDetailsPanel';
import { ASSET_META_DATA_SOURCE } from 'mocks/assets';

storiesOf('AssetDetailsPanel', module)
  .addParameters({
    info: {
      inline: true,
      maxPropObjectKeys: 10,
    },
  })
  .add(
    'Standard',
    () => (
      <AssetDetailsPanel
        dataSource={ASSET_META_DATA_SOURCE}
        columns={AssetDetailsColumns}
      />
    ),
    {
      info: {
        text:
          'Standard view of assetdetailspanel where left-side is the meta-data and right side is the value',
      },
    }
  )
  .add('Empty', () => (
    <AssetDetailsPanel dataSource={[]} columns={AssetDetailsColumns} />
  ))
  .add('No borders', () => (
    <AssetDetailsPanel
      dataSource={ASSET_META_DATA_SOURCE}
      columns={AssetDetailsColumns}
      bordered={false}
    />
  ))
  .add('Show headers', () => (
    <AssetDetailsPanel
      dataSource={ASSET_META_DATA_SOURCE}
      columns={AssetDetailsColumns}
      showHeader={true}
    />
  ))
  .add('Scroll lines', () => (
    <AssetDetailsPanel
      dataSource={ASSET_META_DATA_SOURCE}
      columns={AssetDetailsColumns}
      scroll={{ y: '75vh', x: '100' }}
    />
  ))
  .add('Custom design', () => (
    <AssetDetailsPanel
      dataSource={ASSET_META_DATA_SOURCE}
      columns={AssetDetailsColumns}
      style={{
        border: '4px solid grey',
        margin: '16px',
      }}
    />
  ))
  .add('Repeating same column', () => (
    <AssetDetailsPanel
      dataSource={ASSET_META_DATA_SOURCE}
      columns={[
        ...AssetDetailsColumns,
        ...[{ title: 'name', dataIndex: 'name', key: 'name' }],
      ]}
    />
  ))
  .add('Column with no matching data', () => (
    <AssetDetailsPanel
      dataSource={ASSET_META_DATA_SOURCE}
      columns={[
        ...AssetDetailsColumns,
        ...[{ title: 'nomatch', dataIndex: 'nomatch', key: 'nomatch' }],
      ]}
    />
  ))
  .add('Only one column defined', () => (
    <AssetDetailsPanel
      dataSource={ASSET_META_DATA_SOURCE}
      columns={[{ title: 'name', dataIndex: 'name', key: 'name' }]}
    />
  ));
