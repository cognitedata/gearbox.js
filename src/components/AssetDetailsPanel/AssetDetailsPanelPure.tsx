import React, { FC } from 'react';
import { DescriptionList } from '../DescriptionList';
import { AssetDetailsPanelPureProps } from './interfaces';

export const AssetDetailsPanelPure: FC<AssetDetailsPanelPureProps> = ({
  asset,
  children,
  ...restProps
}) => <DescriptionList valueSet={asset.metadata || {}} {...restProps} />;
