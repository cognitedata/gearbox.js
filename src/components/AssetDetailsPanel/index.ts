export * from './AssetDetailsPanel';
export * from './interfaces';
