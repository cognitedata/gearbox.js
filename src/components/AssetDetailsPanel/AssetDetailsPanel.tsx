import { withAsset } from '../../hoc';
import { AssetDetailsPanelPure } from './AssetDetailsPanelPure';

export const AssetDetailsPanel = withAsset(AssetDetailsPanelPure);
