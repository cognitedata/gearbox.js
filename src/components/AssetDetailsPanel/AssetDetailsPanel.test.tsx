import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { fakeAsset, MockCogniteClient } from '../../mocks';
import { ClientSDKProvider } from '../ClientSDKProvider';
import { LoadingBlock } from '../common/LoadingBlock/LoadingBlock';
import { AssetDetailsPanel } from './AssetDetailsPanel';
import { AssetDetailsPanelPure } from './AssetDetailsPanelPure';
import { AssetDetailsPanelProps } from './interfaces';

configure({ adapter: new Adapter() });

const mockAssetList = jest.fn();

class CogniteClient extends MockCogniteClient {
  assets: any = {
    retrieve: mockAssetList,
  };
}

const sdk = new CogniteClient({ appId: 'gearbox test' });

describe('AssetDetailsPanel', () => {
  beforeEach(() => {
    mockAssetList.mockResolvedValue([fakeAsset]);
  });

  afterEach(() => {
    mockAssetList.mockClear();
  });

  it('Should render without exploding and load data', done => {
    const props: AssetDetailsPanelProps = { assetId: 123 };
    const wrapper = mount(
      <ClientSDKProvider client={sdk}>
        <AssetDetailsPanel {...props} />
      </ClientSDKProvider>
    );
    expect(wrapper.find(LoadingBlock)).toHaveLength(1);

    setImmediate(() => {
      wrapper.update();
      const pureComponent = wrapper.find(AssetDetailsPanelPure);
      expect(pureComponent).toHaveLength(1);
      expect(pureComponent.props().asset).toEqual(fakeAsset);
      done();
    });
  });

  it('Should render categories', done => {
    const props: AssetDetailsPanelProps = {
      assetId: 123,
      toCategory: name => (name.charCodeAt(0) % 3).toString(),
    };
    const wrapper = mount(
      <ClientSDKProvider client={sdk}>
        <AssetDetailsPanel {...props} />
      </ClientSDKProvider>
    );
    setImmediate(() => {
      wrapper.update();
      expect(wrapper.find('.ant-collapse-header').length).toBe(3);
      done();
    });
  });
});
