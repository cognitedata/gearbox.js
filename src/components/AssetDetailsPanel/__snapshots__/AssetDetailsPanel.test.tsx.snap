// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetDetailsPanel renders correctly 1`] = `
<div>
  <div
    className="ant-table-wrapper"
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": "5px",
        "color": "rgba(0, 0, 0, 0.45)",
      }
    }
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-default ant-table-bordered ant-table-without-column-header ant-table-scroll-position-left"
          style={Object {}}
        >
          <div
            className="ant-table-content"
          >
            <div
              className="ant-table-body"
              onScroll={[Function]}
              onWheel={[Function]}
              style={Object {}}
            >
              <table
                className=""
                style={Object {}}
              >
                <colgroup>
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                </colgroup>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="SOURCE_DB"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      SOURCE_DB
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      workmate
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="SOURCE_TABLE"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      SOURCE_TABLE
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      wmate_dba.wmt_location
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="WMT_LOCATION_ID"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      WMT_LOCATION_ID
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      1004
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="WMT_LOCATION_WORKSTART"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      WMT_LOCATION_WORKSTART
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      1999-09-01 07:00:00
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="latestUpdateTimeSource"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      latestUpdateTimeSource
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      1552471210000
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-table-wrapper"
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": "5px",
        "color": "rgba(0, 0, 0, 0.45)",
      }
    }
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-default ant-table-bordered ant-table-empty ant-table-without-column-header ant-table-scroll-position-left"
          style={Object {}}
        >
          <div
            className="ant-table-content"
          >
            <div
              className="ant-table-body"
              onScroll={[Function]}
              onWheel={[Function]}
              style={Object {}}
            >
              <table
                className=""
                style={Object {}}
              >
                <colgroup>
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                </colgroup>
                <tbody
                  className="ant-table-tbody"
                />
              </table>
            </div>
            <div
              className="ant-table-placeholder"
            >
              <div
                className="ant-empty ant-empty-normal"
              >
                <div
                  className="ant-empty-image"
                >
                  <img
                    alt="No Data"
                    src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNjQiIGhlaWdodD0iNDEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CiAgPGcgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoMCAxKSIgZmlsbD0ibm9uZSIgZmlsbC1ydWxlPSJldmVub2RkIj4KICAgIDxlbGxpcHNlIGZpbGw9IiNGNUY1RjUiIGN4PSIzMiIgY3k9IjMzIiByeD0iMzIiIHJ5PSI3Ii8+CiAgICA8ZyBmaWxsLXJ1bGU9Im5vbnplcm8iIHN0cm9rZT0iI0Q5RDlEOSI+CiAgICAgIDxwYXRoIGQ9Ik01NSAxMi43Nkw0NC44NTQgMS4yNThDNDQuMzY3LjQ3NCA0My42NTYgMCA0Mi45MDcgMEgyMS4wOTNjLS43NDkgMC0xLjQ2LjQ3NC0xLjk0NyAxLjI1N0w5IDEyLjc2MVYyMmg0NnYtOS4yNHoiLz4KICAgICAgPHBhdGggZD0iTTQxLjYxMyAxNS45MzFjMC0xLjYwNS45OTQtMi45MyAyLjIyNy0yLjkzMUg1NXYxOC4xMzdDNTUgMzMuMjYgNTMuNjggMzUgNTIuMDUgMzVoLTQwLjFDMTAuMzIgMzUgOSAzMy4yNTkgOSAzMS4xMzdWMTNoMTEuMTZjMS4yMzMgMCAyLjIyNyAxLjMyMyAyLjIyNyAyLjkyOHYuMDIyYzAgMS42MDUgMS4wMDUgMi45MDEgMi4yMzcgMi45MDFoMTQuNzUyYzEuMjMyIDAgMi4yMzctMS4zMDggMi4yMzctMi45MTN2LS4wMDd6IiBmaWxsPSIjRkFGQUZBIi8+CiAgICA8L2c+CiAgPC9nPgo8L3N2Zz4K"
                  />
                </div>
                <p
                  className="ant-empty-description"
                >
                  No Data
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
